File: crates/frf-domain/src/agent.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::ids::{AgentId, SessionId, TenantId};

/// Wire-protocol variant for agent events.
#[non_exhaustive]
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentProtocol {
    AgUi,
    A2a,
    A2ui,
    Custom(String),
}

/// Lifecycle state of an agent event stream.
#[non_exhaustive]
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentEventKind {
    RunStart,
    RunEnd,
    TextDelta,
    ToolCall,
    ToolResult,
    StateSnapshot,
    Error,
    Custom(String),
}

/// Envelope for AG-UI / A2A / A2UI agent protocol events.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AgentEvent {
    pub agent_id: AgentId,
    pub tenant_id: TenantId,
    pub session_id: SessionId,
    pub protocol: AgentProtocol,
    pub kind: AgentEventKind,
    /// Run identifier — groups events in one agent turn.
    pub run_id: String,
    /// Serialized content block (text delta, tool call, etc.).
    pub content: serde_json::Value,
    pub timestamp: DateTime<Utc>,
}

File: admin-ui/src/features/agents/types/agent.ts
export type AgentProtocol = "ag_ui" | "a2a" | "a2ui";

export type AgentEventKind =
  | "run_start"
  | "run_end"
  | "text_delta"
  | "tool_call"
  | "tool_result"
  | "state_snapshot"
  | "error";

export type ContentBlock =
  | { type: "text_delta"; delta: string }
  | { type: "tool_call"; tool_name: string; input: unknown }
  | { type: "tool_result"; tool_name: string; output: unknown; is_error: boolean }
  | { type: "state_snapshot"; state: unknown }
  | { type: "run_start"; model?: string }
  | { type: "run_end"; stop_reason?: string }
  | { type: "error"; message: string; code?: string }
  | { type: string; [key: string]: unknown };

export interface AgentEvent {
  agent_id: string;
  tenant_id: string;
  session_id: string;
  protocol: AgentProtocol;
  kind: AgentEventKind;
  run_id: string;
  content: ContentBlock;
  timestamp: string;
}
